chromium_src/components/omnibox/browser/omnibox_controller.cc

```diff
@@ -5,16 +5,13 @@
 
 #include "components/omnibox/browser/omnibox_controller.h"
 
-#include "brave/components/omnibox/browser/brave_omnibox_client.h"
-
 #define StartAutocomplete StartAutocomplete_ChromiumImpl
 #include "src/components/omnibox/browser/omnibox_controller.cc"
 #undef StartAutocomplete
 
 void OmniboxController::StartAutocomplete(
     const AutocompleteInput& input) const {
-  auto* client = static_cast<BraveOmniboxClient*>(client_.get());
-  if (!client->IsAutocompleteEnabled()) {
+  if (!client_->IsAutocompleteEnabled()) {
```

Comment:
for future reviewers, I suggested Remove BraveOmniboxClient and static_cast to it. That's because there're multiple omnibox clients and I thought it'd better to make it sure all occurrences are guaranteed to    have our new interface (IsAutocompleteEnabled()).

https://github.com/brave/brave-core/pull/22231#discussion_r1502003498

Reply:
FYI, we've got this PR to make a similar change to what's being requested here https://github.com/brave/brave-core/pull/22209. However, it led to an odd crash in my local builds, so we haven't merged it yet. Thought maybe you might find it useful, though.

Reply:
Hahahaha that's nearly identical! Super weird it crashes for you :frowning_face:

Reply:
Thanks @emerick . It happens to me as well

Reply:
Seems like there's a recursive call. I'm looking into it

Reply:
Wow, this is so weird. When I remove the #define macro to add the new method and add the method inline, then it doesn't have crash. Could it be relevant to that omnibox/browser target is static_library?

chromium_src/omnibox_client.h
```c++
// #define CurrentPageExists        \
//   IsAutocompleteEnabled() const; \
//   virtual bool CurrentPageExists
#include "src/components/omnibox/browser/omnibox_client.h"  // IWYU pragma: export
// #undef CurrentPageExists

```


component/omnibox_client.h
```diff
  // Returns whether there is any associated current page.  For example, during
  // startup or shutdown, the omnibox may exist but have no attached page.
  virtual bool CurrentPageExists() const;
+  virtual bool IsAutocompleteEnabled() const;
```

Reply:
That's an interesting possibility about the `static_library` target. But yes, sometimes when I shift things around slightly (in a way that should seemingly have no effect) it "fixes" the crash, but it makes me nervous that I don't know what's going on...

Reply:
That sounds terrifying :laughing: Shall I revert this change, do you reckon?

Reply:
I wonder if someone else form @brave/chromium-src-reviewers could weigh in

Reply:
> That sounds terrifying 😆 Shall I revert this change, do you reckon?

Yeah, sorry @fallaciousreasoning .

Reply:
Let's see what the experts say first :smile:

Reply:
Interestingly, it looks like it's causing failures on the MacOS tests
https://ci.brave.com/job/pr-brave-browser-ntp-search-macos-x64/lastCompletedBuild/testReport/

Reply:
Yeah, it's crashing at startup so all of the Mac tests fail. The stack trace matches what I was seeing with my PR.

Reply:
Interesting - I ran it with `WIPE_WORKSPACE` and CI passed. Would one of you mind checking if a clean build fixes it for you on MacOS?

Reply:
I'm not 100% sure but, this seems to happen to my PR too. https://github.com/brave/brave-core/pull/22304

I rebased it onto the latest and all Web UIs are not loaded. I tried moving code from `chromium_src/../url_data_source.h` to the original file and it was fine. @goodov , do you have any ideas about this?

---

Update, now it's crashing on startup.

Reply:
Yep, I also saw this in my local build yesterday on Mac, and I figured the issue is related to reclient setting `RBE_enable_deps_cache=true` which we have enabled on CI (and I have it enabled locally). In most scenarios it works fine, but when a structure change happens via override that wasn't there before, this can lead to weird recompilation issues when reclient will ignore new dependency and just get the cached version for some object files when in reality they should be recompiled.

On CI this is solvable with `WIPE_WORKSPACE`, because deps cache by default is empty on AMI. `WIPE_WORKSPACE` not just clears workspace, but it recreates the VM from scratch.

Reply:
Fixing it here https://github.com/brave/brave-core/pull/22392

Reply:
Awesome! I removed reproxy_cache dir manually and set RBE_enable_deps=false, and it looks working. Thanks @goodov ! @fallaciousreasoning @emerick , I think we can leave this as is 👍

Reply:
This all makes perfect sense now, thanks everyone!

Reply:
Thanks heaps @goodov!

Reply:
Wow, this change will fix https://github.com/brave/brave-browser/issues/36393 also. 👍🏼 
It crashed because of static casting `RealboxOmniboxClient` to `BraveOmniboxClient`.